mlflow/models/evaluation/default_evaluator.py

```diff
@@ -0,0 +1,399 @@
+
+        try:
+            if model_loader_module == 'mlflow.sklearn':
+                raw_model = model._model_impl
+            elif model_loader_module == 'mlflow.lightgbm':
+                raw_model = model._model_impl.lgb_model
+            elif model_loader_module == 'mlflow.xgboost':
+                raw_model = model._model_impl.xgb_model
+            else:
+                raw_model = None
+        except Exception as e:
+            raw_model = None
+            _logger.warning(f'Raw model resolution fails unexpectedly on PyFuncModel {model!r}, '
+                            f'error message is {e}')
+
+        if raw_model:
+            predict_fn = raw_model.predict
+            try:
+                import xgboost
+                if isinstance(raw_model, xgboost.XGBModel):
+                    # Because shap evaluation will pass evaluation data in ndarray format
+                    # (without feature names), if set validate_features=True it will raise error.
+                    predict_fn = partial(predict_fn, validate_features=False)
+            except ImportError:
+                pass
+
+        sampled_X = shap.sample(X, sample_rows)
+        if algorithm:
+            explainer = shap.Explainer(
+                predict_fn, sampled_X, feature_names=feature_names, algorithm=algorithm
+            )
+            shap_values = explainer(sampled_X)
+        else:
+            maskers = shap.maskers.Independent(sampled_X)
+            if raw_model and shap.explainers.Linear.supports_model_with_masker(raw_model, maskers):
+                explainer = shap.explainers.Linear(
+                    raw_model, maskers, feature_names=feature_names
+                )
+                shap_values = explainer(sampled_X)
+            elif raw_model and shap.explainers.Tree.supports_model_with_masker(raw_model, maskers):
+                explainer = shap.explainers.Tree(
+                    raw_model, maskers, feature_names=feature_names
+                )
+                shap_values = explainer(sampled_X)
+            elif raw_model and shap.explainers.Additive.supports_model_with_masker(
+                    raw_model, maskers
+            ):
+                explainer = shap.explainers.Additive(
+                    raw_model, maskers, feature_names=feature_names
+                )
+                shap_values = explainer(sampled_X)
+            else:
+                # fallback to default sampling explainer
+                explainer = shap.explainers.Sampling(
+                    predict_fn, X, feature_names=feature_names
+                )
+                shap_values = explainer(X, sample_rows)
+
+        _logger.info(f'Shap explainer {explainer.__class__.__name__} is used.')
+
+        # TODO: seems infer pip req fail when log_explainer.
+        mlflow.shap.log_explainer(
+            explainer,
+            artifact_path=DefaultEvaluator._gen_log_key('explainer', dataset_name, model)
+        )
```

Comment:
@WeichenXu123 I'm encountering an error that may or may not be related to your TODO when running the following example:

```
import mlflow
from mlflow.models.evaluation import EvaluationDataset

from sklearn import datasets
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split

# prepare train and test data
iris = datasets.load_iris()
X = iris.data
y = iris.target
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

with mlflow.start_run():
    rid = mlflow.active_run().info.run_id

    logistic = LogisticRegression(C=1.0)
    logistic.fit(X_train, y_train)
    mlflow.sklearn.log_model(logistic, artifact_path="mymodel")

    results = mlflow.evaluate(
        model="runs:/%s/mymodel" % rid,
        model_type="regressor",
        dataset=EvaluationDataset(X_test, y_test),
    )
```

```
/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
2021/12/13 15:24:33 INFO mlflow.models.evaluation.base: Evaluating the model with the default evaluator.
<IPython.core.display.HTML object>
2021/12/13 15:24:34 INFO mlflow.models.evaluation.default_evaluator: Shap explainer Linear is used.
Unable to serialize underlying model using MLflow, will use SHAP serialization
Traceback (most recent call last):
  File "eval.py", line 21, in <module>
    results = mlflow.evaluate(
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/base.py", line 611, in evaluate
    result = evaluator.evaluate(model, model_type, dataset, actual_run_id, config)
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 395, in evaluate
    return self._evaluate_regressor(
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 370, in _evaluate_regressor
    self._log_model_explainability(
  File "/Users/corey.zumar/mlflow/mlflow/models/evaluation/default_evaluator.py", line 211, in _log_model_explainability
    mlflow.shap.log_explainer(
  File "/Users/corey.zumar/mlflow/mlflow/shap.py", line 340, in log_explainer
    Model.log(
  File "/Users/corey.zumar/mlflow/mlflow/models/model.py", line 187, in log
    flavor.save_model(path=local_path, mlflow_model=mlflow_model, **kwargs)
  File "/Users/corey.zumar/mlflow/mlflow/shap.py", line 452, in save_explainer
    explainer.save(explainer_output_file_handle)
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/shap/explainers/_explainer.py", line 407, in save
    s.save("model", self.model, model_saver)
  File "/Users/corey.zumar/opt/anaconda3/lib/python3.8/site-packages/shap/_serializable.py", line 87, in save
    if len(inspect.getfullargspec(value.save)[0]) == 3: # backward compat for MLflow, can remove 4/1/2021
AttributeError: 'LogisticRegression' object has no attribute 'save'
```

Reply:
Reason:
The explainer saver is buggy, if `get_underlying_model_flavor` return "unknown",
then fallback to shap explainer saver, and shap explainer will call `model.save`
for sklearn model, there is no `.save` method, so error will happen.

Reply:
Shall we skip saving shap explainer ? It is easy to cause error.

Reply:
instead of the fallback, can we just do a 'best faith' try except save nothing on the command to log the explainer?

Reply:
Updated shap saving explainer code to be :
```
    with open(explainer_output_path, "wb") as explainer_output_file_handle:
        if serialize_model_using_mlflow and serializable_by_mlflow:
            explainer.save(explainer_output_file_handle, model_saver=False)
        else:
            try:
                explainer.save(explainer_output_file_handle)
            except Exception:
                # Some model may not support explainer model_saver,
                # in this case, fallback to save explainer without saving model.
                explainer_output_file_handle.seek(0)
                explainer_output_file_handle.truncate(0)
                explainer.save(explainer_output_file_handle, model_saver=False)
```

Reply:
@BenWilson2 
But now that, if we skip saving model by `explainer.save(explainer_output_file_handle, model_saver=False)`, when loading back the saved explainer, it will still fail (The shap loader is buggy in this case).

So, shall we only log the explainer class name instead ? If we log a explainer in binary format but it cannot be loaded back, it is useless for end-user

Reply:
Another issue:
Suppose multiple evaluator registered, and each evaluator do log explainer, then it conflicts when logging. Later logged stuff overwrite previous one. Similar things happen for other metric/artifacts. @dbczumar Any thoughts ?